File: XamarinStore.iOS/Views/TopAlignedImageView.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using MonoTouch.UIKit;
using MonoTouch.Foundation;

namespace XamarinStore
{
	public class TopAlignedImageView : UIView
	{
		SizeF origionalSize;
		public UIImage Image
		{
			get { return image; }
			set
			{
				origionalSize = value == null ? SizeF.Empty : value.Size;
				ImageView.Image = image = value;
				LayoutSubviews ();
			}
		}

		UIImageView ImageView;
		UIImage image;
		UIActivityIndicatorView progress;

		public TopAlignedImageView()
		{
			this.ClipsToBounds = true;
			ImageView = new UIImageView();
			this.AddSubview(ImageView);

			AddSubview (progress = new UIActivityIndicatorView (UIActivityIndicatorViewStyle.WhiteLarge));
			this.TranslatesAutoresizingMaskIntoConstraints = false;
		}
		public override void LayoutSubviews()
		{
			base.LayoutSubviews();
			progress.Center = Center;
			if (origionalSize == SizeF.Empty) {
				return;
			}
			var frame = Bounds;
			var scale = frame.Width/origionalSize.Width ;
			frame.Height = origionalSize.Height * scale;
			ImageView.Frame = frame;
		}
		public async void LoadUrl(string url)
		{
			if (string.IsNullOrEmpty (url))
				return;
			var t = FileCache.Download (url);
			if (t.IsCompleted) {
				Image = UIImage.FromFile(t.Result);
				return;
			}
			progress.StartAnimating ();
			var image = UIImage.FromFile(await t);

			UIView.Animate (.3, 
				() => Image = image,
				() => progress.StopAnimating ());
		}

	
	}
}

File: F#/XamarinStore.iOS/Views/TopAlignedImageView.fs
namespace XamarinStore

open System
open System.Collections.Generic
open System.Drawing
open System.Linq
open System.Text
open MonoTouch.UIKit
open MonoTouch.Foundation

open Helpers

type TopAlignedImageView() as this =
    inherit UIView()

    let mutable originalSize = SizeF.Empty
    let mutable image: UIImage = null
    let imageView = new UIImageView()
    let progress = new UIActivityIndicatorView(UIActivityIndicatorViewStyle.WhiteLarge)

    do this.ClipsToBounds <- true
       this.AddSubview imageView
       this.AddSubview progress
       this.TranslatesAutoresizingMaskIntoConstraints <- false

    member this.Image 
        with get () = image
        and set value =
            image <- value
            imageView.Image <- value
            originalSize <- if value = null then SizeF.Empty else value.Size
            this.LayoutSubviews ()

    override this.LayoutSubviews() =
        base.LayoutSubviews()
        progress.Center <- this.Center
        if not (originalSize = SizeF.Empty) then
            let mutable frame = this.Bounds
            let scale = frame.Width/originalSize.Width
            frame.Height <- originalSize.Height * scale
            imageView.Frame <- frame

    member this.LoadUrl(url : string) = async {
        if not (String.IsNullOrEmpty (url)) then
            progress.StartAnimating ()

            let! image = FileCache.Download (url)
            this.Image <- UIImage.FromFile image
            progress.StopAnimating () }

